File: reduxCore/src/main/java/com/develop/zuzik/redux/core/store/reducer/CompositeReducer.java
package com.develop.zuzik.redux.core.store.reducer;

import com.develop.zuzik.redux.core.store.Action;
import com.develop.zuzik.redux.core.store.Reducer;

import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * yaroslavzozulia
 * 10/4/17.
 */

public class CompositeReducer<State> implements Reducer<State> {

	private final List<Reducer<State>> reducers;

	public CompositeReducer(@NotNull List<Reducer<State>> reducers) {
		this.reducers = reducers;
	}

	@NotNull
	@Override
	public State reduce(@NotNull State state, @NotNull Action action) {
		State accumulator = state;
		for (Reducer<State> reducer : reducers) {
			accumulator = reducer.reduce(accumulator, action);
		}
		return accumulator;
	}
}

File: reduxCore/src/main/java/com/develop/zuzik/redux/core/store/reducer/CompositeReducer.kt
package com.develop.zuzik.redux.core.store.reducer

import com.develop.zuzik.redux.core.store.Action
import com.develop.zuzik.redux.core.store.Reducer

/**
 * User: zuzik
 * Date: 4/15/17
 */
class CompositeReducer<State>(private val reducers: List<Reducer<State>>) : Reducer<State> {
	override fun reduce(oldState: State, action: Action): State =
			reducers.fold(oldState) { state, reducer -> reducer.reduce(state, action) }
}